README
The following functions are currently not wrapped and therewith not callable from R. However, 
their wrapping is prepared and commented-out code is present in SnnsCLibWrapper.cpp.
Mostly, only wrapping of some parameters is not implemented. 

Integration of functions beginning with -- probably would require some more work. Integration of all other functions is straightforward,
and was only omitted because they are currently not needed.

-- krui_err     deleteUnitList(int no_of_units, int unit_list[]);
-- krui_err     createFTypeEntry(char *Ftype_symbol, char *act_func_name, char *out_func_name, int no_of_sites, char **array_of_site_names);
int             getFirstPredUnit(FlintType *strength);
int             getFirstPredUnitAndData(FlintType *strength,float *val_a,float *val_b, float *val_c);
int             getNextPredUnit(FlintType *strength);
int             getNextPredUnitAndData(FlintType *strength,float *val_a,float *val_b, float *val_c);
int             getCurrentPredUnit(FlintType *strength);

int             getFirstSuccUnit(int source_unit_no, FlintType *weight);
int             getNextSuccUnit(FlintType *weight);

bool            areConnectedWeight(int source_unit_no,int target_unit_no, FlintType *weight);

struct Link*    createLinkWithAdditionalParameters(int source_unit_no, FlintTypeParam weight,float val_a,float val_b,float val_c);
krui_err        setRemapFunc(char *name, float *params);

-- krui_err     trainNetwork(NetLearnParameters *parameters);
-- krui_err     getNetworkErrorArray(double **learnErrors,int **atEpoch,int *noOfErrors);

krui_err        setClassDistribution(unsigned int *classDist);
krui_err        AlignSubPat(int *inpos, int *outpos, int *no);
krui_err        GetShapeOfSubPattern(int *insize, int *outsize, int *inpos, int *outpos, int n_pos);

-- krui_err     saveResultParam(char *filename, bool create,int startpattern, int endpattern,bool includeinput, bool includeoutput,float *Update_param_array,int NoOfUpdateParam);

krui_err        xyTransTable(int op, int *x, int *y, int z);
krui_err        getUnitCenters(int unit_no, int center_no,struct PositionVector **unit_center);
krui_err        setUnitCenters(int unit_no, int center_no,struct PositionVector *unit_center);

#ifdef MASPAR_KERNEL

krui_err        MasPar( mode );
krui_err        getMasParStatus(void);
krui_err        MasParBenchmark( func_type, cycles, result );

#endif

void            inv_forwardPass(struct UnitList *inputs);
double          inv_backwardPass(float learn, float delta_max,int *err_units, float ratio,struct UnitList *inputs,struct UnitList *outputs);

krui_err        setErrorHandler(void(* error_Handler )(int));



What has been done so far?

-Copy SNNS/kernel/sources to src/
-delete unneeded files: Makefile, kernel.dtt, calc_err.c, make_tbl.c
-create Makevars-file in src/

...


-copy config.h from SNNS/ to src/ 
- remove from config.h relative paths: "kernel/sources/"


-remove from kernel.h functions: kr_searchUnitSiteWithPred (as kernel.h is not included in the swig interface any more, this is probably not necessary)
 and from kr_ui.h: krui_setUnitOutPyFunc


-create RSNNS.i in src/

...

-run swig, move RSNNS.R to R/, replace in RSNNS_wrap.c all true to TRUE and all false to FALSE:

swig -r RSNNS.i
sed -e 's/true/TRUE/g' -e 's/false/FALSE/g' RSNNS_wrap.c > RSNNS_wrap_edt.c
rm RSNNS_wrap.c
mv RSNNS_wrap_edt.c RSNNS_wrap.c
mv RSNNS.R ../R



- For windows compatibility: 
  *run SNNS config script on windows machine (or deactivate some stuff manually)
   in the RTools-shell: 
   sh
   ./configure --build=i386-pc-cygwin32
  *remove alloca stuff from config.h
  *add to config.h: 
  
+#ifndef MAXFLOAT
+#include <float.h>
+#define MAXFLOAT FLT_MAX
+#endif 
  
  *remove alloca.h and alloca.c
  
  *add to kr_pat_scan.c:
    int yywrap() { return 1; }
	
	and comment the YY_SKIP_YYWRAP stuff
  
  *edit lrand48.c: replace "random" by "rand"


-to build swig from svn sources:

AUTOMAKE=automake-1.7 ACLOCAL=aclocal-1.7 ./autogen.sh
./configure --disable-ccache --prefix=/home/bergmeir/Download/swig-build
make
make install


-to compile as cpp:

change ".c" to ".cpp":
for i in *.c; do mv $i $i"pp"; done

//for i in *; do j=`echo $i | cut -d . -f 1`; j=$j"_32.png"; mv $i $j; done
//for i in *.png; do j=`echo $i | cut -d . -f 1`; j=$j".eps"; convert -resize 300x180 $i ./eps/$j; done
//for i in *.pdf; do j=`echo $i | cut -d . -f 1`; j=$j".eps"; convert $i ./eps/$j; done

add some stuff to the .h files:

for i in *.h;
do
sed -e '1i\
#ifdef __cplusplus\
extern "C" {\
#endif' -e '$a\
#ifdef __cplusplus\
}\
#endif' $i > $i"1";

#rm $i;
mv $i"1" $i;
done




change in
/home/bergmeir/Download/swig-build/share/swig/2.0.1/r/r.swg

free($1)
to
//free($1)
otherwise, the *OUTPUT macro of swig is not working



to make automatically generated Rd to work:
sed  -i -e 's/%%  ~~function to do ... ~~/function to do/g' *.Rd


TODOs: 
- extract important things from configure.in of SNNS and build own configure script for 
  the R package (with autoconf), e.g. to determine, if Flex is present


#---------------------------------------

//%apply char* *OUTPUT {char **func_name};                                                        

//%inline %{
//extern void krui_getFuncInfo(int func_no, char **func_name,int *func_type);
//%}

/* Set the input argument to point to a temporary variable */
%typemap(in, numinputs=0) (char **func_name) (char* temp) {
   $1 = &temp;
}

%typemap(argout) (char **func_name) {
   $result = SWIG_AppendResult($result, $1, 1);

}

//
// status is an output parameters for cvCalcOpticalFlowPyrLK
//
%typemap (in, numinputs=1) (char *status) (int tmpCountStatus){
  // as input, we still need the size of the status array

  // memorize the size of the status array
  tmpCountStatus = (int)OctInt_AsLong ($input);

  // create the status array for the C call
  $1 = (char *)malloc (tmpCountStatus * sizeof (char));
}

//
// the status returned by cvCalcOpticalFlowPyrLK
//
%typemap(argout) (char *status) {
  int i;
  octave_value to_add;

  // create the list to return
  to_add = OctList_New (tmpCountStatus$argnum);

  // extract all the integer values of the result, and add it to the
  // final resulting list
  for (i = 0; i < tmpCountStatus$argnum; i++) {
    OctList_SetItem (to_add, i, OctBool_FromLong ($1 [i]));
  }

  $result = SWIG_AppendResult($result, &to_add, 1); 
}



ChangeLog
version 0.4-8 (XX-XX-XXXX):

- single quoted 'RSNNS' in DESCRIPTION (required by CRAN maintainer)
- added a check for NULL in the unit sorting introduced in 0.4-7, to prevent a warning for architectures with no output.
- now checking if the input is numeric, after a conversion to matrix. Before, string input yielded an error in the C++ code and R stopped.
- fixed a bug in the rbfDDA algorithm, related to memory allocation. This will have affected every network architecture that is growing in size of units.
- reimplemented SnnsR__genericPredictCurrPatSet and parts of SnnsR__createPatSet in C++, to get general speed-ups.
- fixed handling of the target variable in DLVQ. Was only working for numeric values starting from 0. Now converting to factor
- added outputActFunc as a parameter to mlp()

version 0.4-7 (05-06-2015):

- fixed a problem occurring with pruned networks, that order of units was not preserved, so that predictions given were wrong (columns changed)
- put NeuralNetTools in suggests field and now mentioning it in the documentation

version 0.4-6 (22-12-2014):

- fixing more memory access issues, discovered by valgrind
- included a bugfix by Tamer Rizk from github, slightly modified
- fixed some "dangling else" warnings
- fixed warnings for misuse of abs() function for floating point values; fabs has to be used in that case
- removed @S3method keywords to fix warnings in roxygen2-4
- now using R's random number generator instead of the system's one

version 0.4-5 (22-05-2014):

- fixing memory access issues, discovered by valgrind and AddressSanitizer
- actually freeing memory now in the destructor
- switch to roxygen2
- elman, jordan: shufflePatterns is now FALSE by default; also changes this in the low-level demo
- included changes of Vitalie Spinu from github
- changed in demos using iris the shuffling to: 
  irisShuffled <- iris[sample(nrow(iris)),]
- fixed bug in iris demo: plotRegressionError was called with targets and fits the wrong way around
- mentioned javaNNS in the documentation
- mentioned the "plot.nnet" function in the documentation
- added some missing wrappers to make pruning work, and included pruning into the mlp function, 
  and two demos for pruning (with the help of Max Greene)
- added imports for methods and Rcpp to NAMESPACE, to comply to new CRAN rules
- fixes for "-Wformat-security", to comply to CRAN rules

version 0.4-4 (04-12-2013):

- fixing a bug for solaris

version 0.4-3 (09-1-2012):

- included a citation to the JSS article about RSNNS in the documentation and the CITATION file
- recompressed example data to fix a warning in CRAN checks
- fixed the jordan and elman examples: "decay=0.1" was given as parameter, instead of the correct "learnFuncParams=c(0.1)" 
- added an example for training an MLP for the iris data, using the low-level api (mlp_irisSnnsR.R)
- fixed the decodeClassLables() function, which didn't correctly translate factor level names to column names  

version 0.4-2 (29-9-2011):

- removed original SNNS manual, to avoid license confusions, as it states an outdated license

version 0.4-1 (28-7-2011):

- fixed a bug that prevented savePatFile to work
- fixed a bug that prevented compilation on Mac OSX
- fixed the "variable set but not used" warnings that occur in the most recent gcc versions

version 0.4-0 (21-6-2011):

- The art2 demo was fixed
- changed package encoding to UTF-8
- wrapping of some of the missing functions was added
- artmap was added in the high-level interface
- initialization to zero was added for all former static variables 
  (this was not always performed before, but it turned out that some
  SNNS functions rely on this and crash otherwise, e.g. dlvq and artmap)
- segfaults related to Rcpp wrapping of (const char*) NULL were fixed
- an object serializaton mechanism was implemented that now allows for rsnns
  objects being saved and loaded through R's normal save/load mechanism
- new normalization functions were added that allow for denormalization,
  and norm of the test set with the parameters obtained on the training set
- some new high-level functions were implemented to extract internals of 
  the neural networks (getWeights, getUnitDefinitions)
- documentation of the high-level functions was greatly improved with references
  to original literature and some descriptions
- original SNNS 4.2 User Manual was added to inst/doc

version 0.3-1 (17-11-2010): 

fixes to make it run on Solaris

version 0.3 (15-11-2010): 

port of the relevant SNNS parts to C++, 
resulting in an SNNS fork named SnnsCLib.

version 0.1 and 0.2 were unpublished. 

version 0.2: 

used nearly unchanged SNNS code. In order to train various networks and 
use them later for prediction, different ways seem possible: After training, 
the model could be saved as a temporary .net file to disk or even to memory 
using e.g. the C function fmemopen. For prediction, the net could be loaded 
into the SNNS kernel again. Another potential possibility is to load the SNNS 
library various times to memory. However, all these methods have mayor drawbacks 
in stability, performance, and are not parallelizable. 

version 0.1: 

used swig instead of Rcpp to wrap the SNNS code. This has the advantage, 
that the wrapper functions are automatically generated. However, it turned 
out that some of the interfaces of SNNS functions are quite complicated 
and the support of swig for R is not as advanced as for other programming 
languages. Extending swig in order to get the needed functionality seemed 
quite difficult, so that we chose to implement the wrapping manually using Rcpp.



R/mlp.R
#############################################################################
#
#   This file is part of the R package "RSNNS".
#
#   Author: Christoph Bergmeir
#   Supervisor: José M. Benítez
#   Copyright (c) DiCITS Lab, Sci2s group, DECSAI, University of Granada.
#
#   This library is free software; you can redistribute it and/or
#   modify it under the terms of the GNU Library General Public
#   License as published by the Free Software Foundation; either
#   version 2 of the License, or (at your option) any later version.
# 
#   This library is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#   Library General Public License for more details.
# 
#   You should have received a copy of the GNU Library General Public License
#   along with this library; see the file COPYING.LIB.  If not, write to
#   the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
#   Boston, MA 02110-1301, USA.
#
#############################################################################


#' This function creates a multilayer perceptron (MLP) and trains it. MLPs are
#' fully connected feedforward networks, and probably the most common network
#' architecture in use.  Training is usually performed by error backpropagation
#' or a related procedure.
#'  
#' There are a lot of different learning functions present in SNNS that can be
#' used together with this function, e.g., \code{Std_Backpropagation},
#' \code{BackpropBatch}, \code{BackpropChunk}, \code{BackpropMomentum},
#' \code{BackpropWeightDecay}, \code{Rprop}, \code{Quickprop}, \code{SCG}
#' (scaled conjugate gradient), ...
#' 
#' \code{Std_Backpropagation}, \code{BackpropBatch}, e.g., have two parameters,
#' the learning rate and the maximum output difference. The learning rate is
#' usually a value between 0.1 and 1. It specifies the gradient descent step
#' width. The maximum difference defines, how much difference between output and
#' target value is treated as zero error, and not backpropagated. This parameter
#' is used to prevent overtraining. For a complete list of the parameters of all
#' the learning functions, see the SNNS User Manual, pp. 67.
#' 
#' The defaults that are set for initialization and update functions usually don't have to be changed. 
#' 
#' 
#' 
#' @title Create and train a multi-layer perceptron (MLP)
#' @references Rosenblatt, F. (1958), 'The perceptron: A probabilistic model for
#' information storage and organization in the brain', Psychological Review
#' 65(6), 386--408.
#' 
#' Rumelhart, D. E.; Clelland, J. L. M. & Group, P. R. (1986), Parallel distributed processing :explorations in the microstructure of cognition, Mit, Cambridge, MA etc.
#'  
#' Zell, A. et al. (1998), 'SNNS Stuttgart Neural Network Simulator User Manual, Version 4.2', IPVR, University of Stuttgart and WSI, University of Tübingen. 
#' \url{http://www.ra.cs.uni-tuebingen.de/SNNS/}
#' 
#' Zell, A. (1994), Simulation Neuronaler Netze, Addison-Wesley. (in German)
#' @export
mlp <- function(x, ...) UseMethod("mlp")


#' @param x a matrix with training inputs for the network
#' @param y the corresponding targets values
#' @param size number of units in the hidden layer(s)
#' @param maxit maximum of iterations to learn
#' @param initFunc the initialization function to use
#' @param initFuncParams the parameters for the initialization function
#' @param learnFunc the learning function to use
#' @param learnFuncParams the parameters for the learning function
#' @param updateFunc the update function to use
#' @param updateFuncParams the parameters for the update function
#' @param hiddenActFunc the activation function of all hidden units
#' @param shufflePatterns should the patterns be shuffled?
#' @param linOut sets the activation function of the output units to linear or logistic
#' @param inputsTest a matrix with inputs to test the network
#' @param targetsTest the corresponding targets for the test input
#' @param pruneFunc the pruning function to use
#' @param pruneFuncParams the parameters for the pruning function. Unlike the other functions, 
#' these have to be given in a named list. See the pruning demos for further explanation. 
#' @param ... additional function parameters (currently not used)
#' @return an \code{\link{rsnns}} object.
#' @export
#' @S3method mlp default
#' @method mlp default
#' @rdname mlp
#' @examples 
#' \dontrun{demo(iris)}
#' \dontrun{demo(laser)}
#' \dontrun{demo(encoderSnnsCLib)}
#' 
#' 
#' data(iris)
#' 
#' #shuffle the vector
#' iris <- iris[sample(1:nrow(iris),length(1:nrow(iris))),1:ncol(iris)]
#' 
#' irisValues <- iris[,1:4]
#' irisTargets <- decodeClassLabels(iris[,5])
#' #irisTargets <- decodeClassLabels(iris[,5], valTrue=0.9, valFalse=0.1)
#' 
#' iris <- splitForTrainingAndTest(irisValues, irisTargets, ratio=0.15)
#' iris <- normTrainingAndTestSet(iris)
#' 
#' model <- mlp(iris$inputsTrain, iris$targetsTrain, size=5, learnFuncParams=c(0.1), 
#'               maxit=50, inputsTest=iris$inputsTest, targetsTest=iris$targetsTest)
#' 
#' summary(model)
#' model
#' weightMatrix(model)
#' extractNetInfo(model)
#' 
#' par(mfrow=c(2,2))
#' plotIterativeError(model)
#' 
#' predictions <- predict(model,iris$inputsTest)
#' 
#' plotRegressionError(predictions[,2], iris$targetsTest[,2])
#' 
#' confusionMatrix(iris$targetsTrain,fitted.values(model))
#' confusionMatrix(iris$targetsTest,predictions)
#' 
#' plotROC(fitted.values(model)[,2], iris$targetsTrain[,2])
#' plotROC(predictions[,2], iris$targetsTest[,2])
#' 
#' #confusion matrix with 402040-method
#' confusionMatrix(iris$targetsTrain, encodeClassLabels(fitted.values(model),
#'                                                        method="402040", l=0.4, h=0.6))
mlp.default <- function(x, y, size=c(5), maxit=100,  
    initFunc="Randomize_Weights", initFuncParams=c(-0.3, 0.3), 
    learnFunc="Std_Backpropagation", learnFuncParams=c(0.2, 0.0), 
    updateFunc="Topological_Order", updateFuncParams=c(0.0),
    hiddenActFunc="Act_Logistic",
    shufflePatterns=TRUE, linOut=FALSE, inputsTest=NULL, targetsTest=NULL, pruneFunc=NULL, pruneFuncParams=NULL, ...) {


  x <- as.matrix(x)
  y <- as.matrix(y)

  checkInput(x, y)

  nInputs <- ncol(x)
  nOutputs <- ncol(y)


  snns <- rsnnsObjectFactory(
    subclass=c("mlp"), nInputs=nInputs, maxit=maxit, 
    initFunc=initFunc, initFuncParams=initFuncParams, 
    learnFunc=learnFunc, learnFuncParams=learnFuncParams, 
    updateFunc=updateFunc, updateFuncParams=updateFuncParams,
    shufflePatterns=shufflePatterns, computeIterativeError=TRUE,
    pruneFunc=pruneFunc, pruneFuncParams=pruneFuncParams)
  
  snns$archParams <- list(size=size)
  snns$snnsObject$setUnitDefaults(0,0,1,0,1,"Act_Logistic","Out_Identity")
  snns$snnsObject$createNet(unitsPerLayer=c(nInputs, size, nOutputs),
                            fullyConnectedFeedForward=TRUE,
                            iNames = colnames(x), oNames = colnames(y))
  
  outputActFunc <- if(linOut)  "Act_Identity" else  "Act_Logistic"
  
  snns$snnsObject$setTTypeUnitsActFunc("UNIT_INPUT", "Act_Identity")
  snns$snnsObject$setTTypeUnitsActFunc("UNIT_HIDDEN", hiddenActFunc)
  snns$snnsObject$setTTypeUnitsActFunc("UNIT_OUTPUT", outputActFunc)
  
  snns <- train(snns,
                inputsTrain = x, inputsTest = inputsTest,
                targetsTrain = y, targetsTest = targetsTest)
  
}

